core/node/consensus/src/storage/vm_reader.rs

```diff
@@ -0,0 +1,341 @@
+            tokens[2].clone().into_bool().unwrap(),
+        )
+    }
+
+    async fn read_committee_validator(
+        &mut self,
+        ctx: &Ctx,
+        block_id: BlockId,
+        idx: usize,
+    ) -> (Address, usize, Vec<u8>) {
+        let func = self
+            .validator_registry_contract
+            .function("committee")
+            .unwrap()
+            .clone();
+        let tx = self.gen_l2_call_tx(
+            self.validator_registry_address.unwrap(),
+            func.encode_input(&[Token::Uint(zksync_types::U256::from(idx))])
+                .unwrap(),
+        );
+
+        let res = self.eth_call(ctx, block_id, tx).await;
+        let tokens = func.decode_output(&res).unwrap();
+        (
+            tokens[0].clone().into_address().unwrap(),
+            tokens[1].clone().into_uint().unwrap().as_usize(),
+            tokens[2].clone().into_bytes().unwrap(),
+        )
+    }
+
+    async fn read_committee_attester(
+        &mut self,
+        ctx: &Ctx,
+        block_id: BlockId,
+        idx: usize,
+    ) -> (usize, Address, Vec<u8>) {
+        let func = self
+            .attester_registry_contract
+            .function("committee")
+            .unwrap()
+            .clone();
+        let tx = self.gen_l2_call_tx(
+            self.attester_registry_address.unwrap(),
+            func.encode_input(&[Token::Uint(zksync_types::U256::from(idx))])
+                .unwrap(),
+        );
+
+        let res = self.eth_call(ctx, block_id, tx).await;
+        let tokens = func.decode_output(&res).unwrap();
+        (
+            tokens[0].clone().into_uint().unwrap().as_usize(),
+            tokens[1].clone().into_address().unwrap(),
+            tokens[2].clone().into_bytes().unwrap(),
+        )
+    }
+
+    async fn read_address(
+        &mut self,
+        ctx: &Ctx,
+        block_id: BlockId,
+        contract_address: Address,
+        func: Function,
+    ) -> Address {
+        let tx = self.gen_l2_call_tx(contract_address, func.encode_input(&vec![]).unwrap());
+
+        let res = self.eth_call(ctx, block_id, tx).await;
+        let tokens = func.decode_output(&res).unwrap();
+        tokens[0].clone().into_address().unwrap()
+    }
+
+    async fn eth_call(&mut self, ctx: &Ctx, block_id: BlockId, tx: L2Tx) -> Vec<u8> {
+        let mut conn = self.pool.connection(ctx).await.unwrap().0;
+        let start_info = BlockStartInfo::new(&mut conn, Duration::from_secs(10))
+            .await
+            .unwrap();
+        let block_args = zksync_node_api_server::execution_sandbox::BlockArgs::new(
+            &mut conn,
+            block_id,
+            &start_info,
+        )
+        .await
+        .unwrap();
+        let call_overrides = CallOverrides {
+            enforced_base_fee: None,
+        };
+
+        let res = self
+            .tx_sender
+            .eth_call(block_args, call_overrides, tx)
+            .await
+            .unwrap();
```

Comment:
If we want to consider running VM to query contracts (as opposed to pre-calculating storage slots and reading storage directly), I think we need to do quite some preparation work:

- Extract `execution_vm` into a separate, general-purpose crate.
- Change the sandbox so that it can perform multiple requests on a single instance.
- Decouple it from `TxSender`.

Right now, `tx_sender` and dependency on `zksync_node_api_server` feel like a pretty huge hack, and I'm not sure if it's a good idea to depend on that.

Reply:
@popzxc In your opinion, should we read the storage slots directly instead of using VM read transactions?

Reply:
I can certainly see pros of using VM with `eth_call`, but I believe that implementing it properly would take _a lot_ of engineering efforts. As you can see, the first try is implemented in a hacky way, and making a _proper_ implementation (e.g. detaching sandbox from API) is a complex task even for seasoned platform developers. I would really like to not use hacky ways, as they may backfire, especially given that consensus will be a critical component for the sequencer.

On the other hand, reading storage directly will never look as "good", but it's relatively easy to implement and cover with unit tests (e.g. change value via contract method, then read from storage directly) to make sure that the storage layout doesn't change.

If we can afford to spend time on extracting the sandbox from API, I'm all for it. If not, I would rather prefer direct storage access than hacks.

Reply:
I agree that the current solution is a workaround/hack as per the code dependency graph and the lack of execution sandbox instance re-use.

However, falling back to direct storage reads is worse IMHO.

The reads are not targeting U256 variables in which their storage slots’ keys can be easily pre-calculated, be fetched and decoded. It involves dynamic-sized arrays, structs, and dynamic-sized arrays within the structs. This implies packing/padding, which means to re-implement the VM codec. 

Meanwhile, I couldn’t find any relevant documentation, and there’s very limited code examples (mostly in `zksync_types::utils`), for much simpler use-cases. 

I went ahead to experiment with it, and couldn’t even retrieve a simple U256 variable slot, following canonical `solc` storage layout. I might have overlooked something, but couldn’t reverse-engineer the slots back to the pre-hashed keys. In the overall, I don’t think it’s a good idea to go forward with this given the complex data types involved. 

Also, I don’t think such approach is testable, because the layout is supposedly an implementation detail, unless there’s a clear open available specs, backwards compatibility, and lack of potential optimizations. 

Read-transactions are the standard way to read from the VM storage database, while incapsulating the layout and the codec. The consensus component functions as a standard web3 client in this regard, and having direct access to the database does not imply it should use it to bypass the standard interface. 

I suggest to continue with the current approach, and to implement your suggestions for the execution sandbox decoupling, in a follow-up PR. (see BFT-493)

Reply:
+1

Reply:
I don't mind as long as refactoring will really be planned & executed.